app/models/user.rb

```diff
@@ -1,11 +1,22 @@
 class User < ApplicationRecord
-  before_save { self.email = email.downcase }
-  validates :name, presence: true, length: { maximum: 50 }
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-  validates :email, presence: true, length: { maximum: 255 },
-            format: { with: VALID_EMAIL_REGEX },
-            uniqueness: { case_sensitive: false }
 
+  validates :name, presence: true,
+    length: {maximum: Setting.name_max_length}
+  validates :email, presence: true,
+    length: {maximum: Setting.email_max_length},
+    format: {with: VALID_EMAIL_REGEX},
+    uniqueness: {case_sensitive: false}
+  validates :password, presence: true,
+    length: {minimum: Setting.password_min_length}
 
+  has_secure_password
```

Comment:
cái này sẽ chạy những logic nào em

Reply:
has_secure_password là một module của gem "bcrypt", cụ thể hơn là module của ActiveModel::SecurePassword::ClassMethods. Khi thêm phương thức has_secure_password thuộc module này, nó cung cấp để định nghĩa một mật khẩu xác nhận hợp lệ. Nó có chức năng nhu:
Mật khẩu phải tồn tại.
Mật khẩu phải khớp với phần xác nhận mật khẩu.
Mật khẩu không được vượt quá 72 ký tự.

Ở trong chappter này thì has_secure_password có khả năng lưu thuộc tính password_digest được băm an toàn vào cơ sở dữ liệu Một cặp thuộc tính ảo (mật khẩu và mật khẩu xác nhận), bao gồm xác thực hiện diện khi tạo đối tượng và xác thực yêu cầu chúng khớp với nhau. Một phương thức xác thực trả về người dùng khi mật khẩu đúng (hoặc sai).
has_secure_password thực hiện xác thực  the virtual password and password_confirmation attributes. Khi người dùng gửi 1 request lên gồm password và password_confirmation lên, nếu cả 2 bằng nhau thì nó sẽ tạo 1 password_digest lưu vào db (Tất nhiên ở đây phải đảm bảo là gmail phải đúng và 1 số điều kiện phải thoả mãn thì mới thực hiện create user được)